assets/js/modules/analytics-4/datastore/custom-dimensions-gathering-data.js

```diff
@@ -0,0 +1,395 @@
+		invariant(
+			'string' === typeof customDimension,
+			'customDimension must be a string.'
+		);
+		invariant(
+			'boolean' === typeof gatheringData,
+			'gatheringData must be a boolean.'
+		);
+
+		return {
+			payload: {
+				customDimension,
+				gatheringData,
+			},
+			type: RECEIVE_CUSTOM_DIMENSION_GATHERING_DATA,
+		};
+	},
+
+	/**
+	 * Receives data available on load state for a custom dimension.
+	 *
+	 * This action was added to easily manipulate the state for
+	 * JS tests and Storybook / VRT scenarios.
+	 *
+	 * @since n.e.x.t
+	 * @private
+	 *
+	 * @param {string}  customDimension     Custom dimension slug.
+	 * @param {boolean} dataAvailableOnLoad Data availability on load.
+	 * @return {Object} Redux-style action.
+	 */
+	receiveIsCustomDimensionDataAvailableOnLoad(
+		customDimension,
+		dataAvailableOnLoad
+	) {
+		invariant(
+			'string' === typeof customDimension,
+			'customDimension must be a string.'
+		);
+		invariant(
+			'boolean' === typeof dataAvailableOnLoad,
+			'dataAvailableOnLoad must be a boolean.'
+		);
+
+		return {
+			payload: {
+				customDimension,
+				dataAvailableOnLoad,
+			},
+			type: RECEIVE_CUSTOM_DIMENSION_DATA_AVAILABLE_ON_LOAD,
+		};
+	},
+};
+
+const baseControls = {
+	[ WAIT_FOR_CUSTOM_DIMENSION_DATA_AVAILABILITY_STATE ]:
+		createRegistryControl(
+			( registry ) =>
+				( { payload: { customDimension } } ) => {
+					const dataAvailabityDetermined = () =>
+						registry
+							.select( MODULES_ANALYTICS_4 )
+							.selectCustomDimensionDataAvailability(
+								customDimension
+							) !== undefined;
+
+					if ( dataAvailabityDetermined() ) {
+						return true;
+					}
+
+					return new Promise( ( resolve ) => {
+						const unsubscribe = registry.subscribe( () => {
+							if ( dataAvailabityDetermined() ) {
+								unsubscribe();
+								resolve( true );
+							}
+						} );
+					} );
+				}
+		),
+};
+
+const baseReducer = createReducer( ( state, { type, payload } ) => {
+	switch ( type ) {
+		case RECEIVE_CUSTOM_DIMENSION_GATHERING_DATA: {
+			const { customDimension, gatheringData } = payload;
+
+			state.customDimensionsGatheringData[ customDimension ] =
+				gatheringData;
+
+			break;
+		}
+
+		case RECEIVE_CUSTOM_DIMENSION_DATA_AVAILABLE_ON_LOAD: {
+			const { customDimension, dataAvailableOnLoad } = payload;
+
+			state.customDimensionsDataAvailableOnLoad[ customDimension ] =
+				dataAvailableOnLoad;
+
+			break;
+		}
+
+		default: {
+			break;
+		}
+	}
+} );
+
+const baseResolvers = {
+	*isCustomDimensionGatheringData( customDimension ) {
+		const registry = yield Data.commonActions.getRegistry();
+
+		// If the gatheringData flag is already set, return early.
+		if (
+			registry
+				.select( MODULES_ANALYTICS_4 )
+				.isCustomDimensionGatheringData( customDimension ) !== undefined
+		) {
+			return;
+		}
+
+		const dataAvailableOnLoad = registry
+			.select( MODULES_ANALYTICS_4 )
+			.isCustomDimensionDataAvailableOnLoad( customDimension );
```

Comment:
This seems like an unnecessary selector. See how we do a resolver for site info: https://github.com/google/site-kit-wp/blob/5aafc62bfa9449790264593dba048da530ee54a7/assets/js/googlesitekit/datastore/site/info.js#L236-L314

We should use the same approach here for preloaded data. It will let us remove a bunch of redundant stuff from this file that is currently used to manage preloaded data.

Reply:
That's a good point. I had based this implementation quite closely on `createGatheringDataStore()` but you're right, we can do without a good chunk of it. I've simplified accordingly and removed the unnecessary code.